src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10949,99 +10946,144 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
             // parse out a missing name node for the expression, and keep on going
             var tk = this.CurrentToken.Kind;
             if (IsInvalidSubExpression(tk))
-            {
                 return this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
-            }
 
-            // Parse a left operand -- possibly preceded by a unary operator.
-            if (IsExpectedPrefixUnaryOperator(tk))
-            {
-                var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
-                newPrecedence = GetPrecedence(opKind);
-                var opToken = this.EatToken();
-                var operand = this.ParseSubExpression(newPrecedence);
-                leftOperand = _syntaxFactory.PrefixUnaryExpression(opKind, opToken, operand);
-            }
-            else if (tk == SyntaxKind.DotDotToken)
+            return ParseExpressionContinued(parseUnaryOrPrimaryExpression(precedence), precedence);
+
+            // Parses out a unary expression, or something with higher precedence (cast, addressof, primary).
+            ExpressionSyntax parseUnaryOrPrimaryExpression(Precedence precedence)
             {
-                // Operator ".." here can either be a prefix unary operator or a stand alone empty range:
-                var opToken = this.EatToken();
-                newPrecedence = GetPrecedence(SyntaxKind.RangeExpression);
+                // Parse a left operand -- possibly preceded by a unary operator.
+                if (IsExpectedPrefixUnaryOperator(tk))
+                {
+                    var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
+                    return _syntaxFactory.PrefixUnaryExpression(
+                        opKind,
+                        this.EatToken(),
+                        this.ParseSubExpression(GetPrecedence(opKind)));
+                }
 
-                ExpressionSyntax rightOperand;
-                if (CanStartExpression())
+                if (tk == SyntaxKind.DotDotToken)
                 {
-                    rightOperand = this.ParseSubExpression(newPrecedence);
+                    return _syntaxFactory.RangeExpression(
+                        leftOperand: null,
+                        this.EatToken(),
+                        CanStartExpression()
+                            ? this.ParseSubExpression(Precedence.Range)
+                            : null);
```

Comment:
example of simplifying.  instead of computing the nested precedence dynamically (even though it is fixed), and assigning to an outer varaible (With lack of clarity who is affected by that), we simply parse the subexpression with the correct precedence for this construct.

Reply:
note: i'm fairly certain we don't need this.  a range expression is not a primary or unary expression.  this should get picked up in teh 'continur parsing' routine.  will try that out.

Reply:
by naming the functions appropriately, this becomes clearer.

Reply:
> this should get picked up in teh 'continur parsing' routine. will try that out.

Interested in finding out the result of this.

Reply:
answered in main thread.  and doc'ed here.